File: client/src/lib/levelGenerators/shadow.ts
import { Point, clamp, indexFor } from "./shared";
import { mulberry32 as rng } from "../rng";

export type ShadowLevel = {
  size: number;
  pads: Point[];
  exit: Point;
  inverseTiles: Point[];
  lag: number;
  lesson: string;
  lessonEn?: string;
};

/**
 * Gölge Payı (Shadow) seviye üreteci:
 * Oyuncu ile gecikmeli gölge (lag: 2 veya 3 adım) koordineli olarak tahtadaki iki baskı pedine
 * aynı anda basmalıdır. Ped B, Ped A'dan tam olarak `lag` adımda erişilebilecek konumlara
 * yerleştirilir. Böylece oyuncu bir pedden diğerine kesintisiz yürüdüğünde gölge ilk pede varır
 * ve kapı doğal olarak açılır (yapay bekleme hamlesi gerekmez). Çözücü (BFS) 4 yönde seviyenin
 * 12 denemede kesinlikle kazanılabilir olduğunu doğrular.
 */
function buildShadowLevelCandidate(seed: number, mastery: number): ShadowLevel {
  const size = mastery >= 3 ? 6 : 5;
  const random = rng(seed ^ 0x5bd1e995);
  const lag = mastery >= 4 ? 3 : 2;

  // Çıkış noktası kenar boyunca seçilir
  const exitCandidates: Point[] = [];
  for (let i = 1; i < size; i += 1) {
    exitCandidates.push({ x: size - 1, y: i });
    exitCandidates.push({ x: i, y: size - 1 });
  }
  const exit = exitCandidates[indexFor(seed, 211, exitCandidates.length)];

  const inner = () => 1 + Math.floor(random() * (size - 2));
  const colA = inner();
  const rowA = inner();

  // Ped B: lag mesafesinde erişilebilir noktalar
  let offsets: Array<{ dx: number; dy: number }> = [];
  if (lag === 2) {
    offsets = [
      { dx: 2, dy: 0 }, { dx: -2, dy: 0 }, { dx: 0, dy: 2 }, { dx: 0, dy: -2 },
      { dx: 1, dy: 1 }, { dx: 1, dy: -1 }, { dx: -1, dy: 1 }, { dx: -1, dy: -1 },
    ];
  } else {
    // lag === 3
    offsets = [
      { dx: 1, dy: 0 }, { dx: -1, dy: 0 }, { dx: 0, dy: 1 }, { dx: 0, dy: -1 },
      { dx: 3, dy: 0 }, { dx: -3, dy: 0 }, { dx: 0, dy: 3 }, { dx: 0, dy: -3 },
      { dx: 2, dy: 1 }, { dx: 2, dy: -1 }, { dx: -2, dy: 1 }, { dx: -2, dy: -1 },
      { dx: 1, dy: 2 }, { dx: 1, dy: -2 }, { dx: -1, dy: 2 }, { dx: -1, dy: -2 },
    ];
  }

  const validPadsB: Point[] = [];
  for (const off of offsets) {
    const bx = colA + off.dx;
    const by = rowA + off.dy;
    if (bx >= 0 && bx < size && by >= 0 && by < size) {
      if ((bx !== 0 || by !== 0) && (bx !== exit.x || by !== exit.y) && (bx !== colA || by !== rowA)) {
        validPadsB.push({ x: bx, y: by });
      }
    }
  }

  const padB = validPadsB.length > 0
    ? validPadsB[indexFor(seed, 937, validPadsB.length)]
    : { x: colA, y: clamp(rowA + (lag === 2 ? 2 : 1), 0, size - 1) };

  const inverseTiles: Point[] = mastery >= 3 ? [{ x: inner(), y: inner() }] : [];

  const lesson = mastery >= 3
    ? "Işık prizmasından geçen gölgenin yönü tersine döner. Gecikmeli gölgeni arkandan sürükleyerek iki pedi aynı anda aktif et."
    : `Gölgen ${lag} hamle geriden gelir. İki pedi eşleştirmek için bir pedden diğerine ${lag} adımda kesintisiz geç.`;

  const lessonEn = mastery >= 3
    ? "Crossing light prisms inverts the shadow's direction. Trace your steps so your delayed shadow matches both pads."
    : `Your shadow follows ${lag} steps behind. Walk from one pad to the other in exactly ${lag} steps to match both.`;

  return {
    size,
    pads: [{ x: colA, y: rowA }, padB],
    exit,
    inverseTiles,
    lag,
    lesson,
    lessonEn,
  };
}

export function generateShadowLevel(seed: number, mastery: number): ShadowLevel {
  for (let attempt = 0; attempt < 12; attempt += 1) {
    const candidate = buildShadowLevelCandidate(seed + attempt * 7331, mastery);
    if (isShadowLevelSolvable(candidate)) return candidate;
  }
  return buildShadowLevelCandidate(seed, mastery);
}

export function isShadowLevelSolvable(level: ShadowLevel) {
  type ShadowPosition = { r: number; c: number };
  type State = { player: ShadowPosition; shadow: ShadowPosition; history: Array<[number, number]>; open: boolean };
  const at = (position: ShadowPosition, point: Point) => position.c === point.x && position.r === point.y;
  const onPad = (position: ShadowPosition) => level.pads.some(point => at(position, point));
  const inverse = (position: ShadowPosition) => level.inverseTiles.some(point => at(position, point));
  const keyFor = (state: State) => `${state.player.r},${state.player.c}/${state.shadow.r},${state.shadow.c}/${state.history.map(move => move.join(":")).join("|")}/${state.open ? 1 : 0}`;
  const queue: Array<State & { steps: number }> = [{ player: { r: 0, c: 0 }, shadow: { r: 0, c: 0 }, history: [], open: false, steps: 0 }];
  const seen = new Set([keyFor(queue[0])]);

  // Sadece 4 yön (doğal ızgara hareketi, bekleme yok)
  const moves: Array<[number, number]> = [[-1, 0], [1, 0], [0, -1], [0, 1]];

  while (queue.length) {
    const current = queue.shift()!;
    if (current.open && at(current.player, level.exit)) return true;
    if (current.steps >= 36) continue;

    for (const [dr, dc] of moves) {
      const player = {
        r: clamp(current.player.r + dr, 0, level.size - 1),
        c: clamp(current.player.c + dc, 0, level.size - 1),
      };

      if (player.r === current.player.r && player.c === current.player.c) continue;

      let shadow = current.shadow;
      if (current.history.length >= level.lag) {
        const [lagDr, lagDc] = current.history[0];
        const factor = inverse(current.shadow) ? -1 : 1;
        shadow = {
          r: clamp(current.shadow.r + lagDr * factor, 0, level.size - 1),
          c: clamp(current.shadow.c + lagDc * factor, 0, level.size - 1),
        };
      }

      const history = [...current.history, [dr, dc] as [number, number]].slice(-level.lag);
      const open = current.open || (onPad(player) && onPad(shadow) && (player.r !== shadow.r || player.c !== shadow.c));
      const next = { player, shadow, history, open, steps: current.steps + 1 };
      const key = keyFor(next);
      if (!seen.has(key)) {
        seen.add(key);
        queue.push(next);
      }
    }
  }
  return false;
}

File: src/games/shadow.rs
//! Shadow game level generation and solver
//! Matches client/src/lib/levelGenerators/shadow.ts.

use std::collections::{HashSet, VecDeque};
use serde::{Deserialize, Serialize};
use super::rng::Mulberry32;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShadowPoint {
    pub x: i32,
    pub y: i32,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ShadowLevel {
    pub size: usize,
    pub pads: Vec<ShadowPoint>,
    pub exit: ShadowPoint,
    pub inverse_tiles: Vec<ShadowPoint>,
    pub lag: usize,
    pub lesson: String,
    pub lesson_en: Option<String>,
}

pub fn index_for(seed: u32, salt: u32, length: usize) -> usize {
    if length == 0 {
        return 0;
    }
    let val = (seed.wrapping_add(salt)).wrapping_mul(1103515245) as i32;
    ((val as i64).abs() as usize) % length
}

fn clamp(val: i32, min: i32, max: i32) -> i32 {
    val.max(min).min(max)
}

pub fn build_shadow_level_candidate(seed: u32, mastery: u32) -> ShadowLevel {
    let size = if mastery >= 3 { 6 } else { 5 };
    let mut random = Mulberry32::new(seed ^ 0x5bd1e995);
    let lag = if mastery >= 4 { 3 } else { 2 };

    let mut exit_candidates: Vec<ShadowPoint> = Vec::new();
    for i in 1..size {
        exit_candidates.push(ShadowPoint { x: size as i32 - 1, y: i as i32 });
        exit_candidates.push(ShadowPoint { x: i as i32, y: size as i32 - 1 });
    }
    let exit = exit_candidates[index_for(seed, 211, exit_candidates.len())];

    let mut inner = || 1 + (random.next_f64() * (size - 2) as f64).floor() as i32;
    let col_a = inner();
    let row_a = inner();

    let offsets: Vec<(i32, i32)> = if lag == 2 {
        vec![
            (2, 0), (-2, 0), (0, 2), (0, -2),
            (1, 1), (1, -1), (-1, 1), (-1, -1),
        ]
    } else {
        vec![
            (1, 0), (-1, 0), (0, 1), (0, -1),
            (3, 0), (-3, 0), (0, 3), (0, -3),
            (2, 1), (2, -1), (-2, 1), (-2, -1),
            (1, 2), (1, -2), (-1, 2), (-1, -2),
        ]
    };

    let mut valid_pads_b: Vec<ShadowPoint> = Vec::new();
    for (dx, dy) in offsets {
        let bx = col_a + dx;
        let by = row_a + dy;
        let s = size as i32;
        if bx >= 0 && bx < s && by >= 0 && by < s {
            if (bx != 0 || by != 0) && (bx != exit.x || by != exit.y) && (bx != col_a || by != row_a) {
                valid_pads_b.push(ShadowPoint { x: bx, y: by });
            }
        }
    }

    let pad_b = if !valid_pads_b.is_empty() {
        valid_pads_b[index_for(seed, 937, valid_pads_b.len())]
    } else {
        ShadowPoint {
            x: col_a,
            y: clamp(row_a + if lag == 2 { 2 } else { 1 }, 0, size as i32 - 1),
        }
    };

    let inverse_tiles: Vec<ShadowPoint> = if mastery >= 3 {
        vec![ShadowPoint { x: inner(), y: inner() }]
    } else {
        Vec::new()
    };

    let lesson = if mastery >= 3 {
        "Işık prizmasından geçen gölgenin yönü tersine döner. Gecikmeli gölgeni arkandan sürükleyerek iki pedi aynı anda aktif et.".to_string()
    } else {
        format!(
            "Gölgen {} hamle geriden gelir. İki pedi eşleştirmek için bir pedden diğerine {} adımda kesintisiz geç.",
            lag, lag
        )
    };

    let lesson_en = if mastery >= 3 {
        Some("Crossing light prisms inverts the shadow's direction. Trace your steps so your delayed shadow matches both pads.".to_string())
    } else {
        Some(format!(
            "Your shadow follows {} steps behind. Walk from one pad to the other in exactly {} steps to match both.",
            lag, lag
        ))
    };

    ShadowLevel {
        size,
        pads: vec![ShadowPoint { x: col_a, y: row_a }, pad_b],
        exit,
        inverse_tiles,
        lag,
        lesson,
        lesson_en,
    }
}

pub fn is_shadow_level_solvable(level: &ShadowLevel) -> bool {
    let size = level.size as i32;

    let on_pad = |r: i32, c: i32| -> bool {
        level.pads.iter().any(|p| p.x == c && p.y == r)
    };

    let is_inverse = |r: i32, c: i32| -> bool {
        level.inverse_tiles.iter().any(|p| p.x == c && p.y == r)
    };

    #[derive(Clone, PartialEq, Eq, Hash)]
    struct StateKey {
        player: (i32, i32),
        shadow: (i32, i32),
        history: Vec<(i32, i32)>,
        open: bool,
    }

    struct QueueItem {
        state: StateKey,
        steps: usize,
    }

    let initial_key = StateKey {
        player: (0, 0),
        shadow: (0, 0),
        history: Vec::new(),
        open: false,
    };

    let mut seen: HashSet<StateKey> = HashSet::new();
    seen.insert(initial_key.clone());

    let mut queue: VecDeque<QueueItem> = VecDeque::new();
    queue.push_back(QueueItem {
        state: initial_key,
        steps: 0,
    });

    let moves: [(i32, i32); 4] = [(-1, 0), (1, 0), (0, -1), (0, 1)];

    while let Some(current) = queue.pop_front() {
        if current.state.open && current.state.player.0 == level.exit.y && current.state.player.1 == level.exit.x {
            return true;
        }

        if current.steps >= 36 {
            continue;
        }

        for &(dr, dc) in &moves {
            let next_player = (
                clamp(current.state.player.0 + dr, 0, size - 1),
                clamp(current.state.player.1 + dc, 0, size - 1),
            );

            if next_player == current.state.player {
                continue;
            }

            let mut next_shadow = current.state.shadow;
            if current.state.history.len() >= level.lag {
                let (lag_dr, lag_dc) = current.state.history[0];
                let factor = if is_inverse(next_shadow.0, next_shadow.1) { -1 } else { 1 };
                next_shadow = (
                    clamp(next_shadow.0 + lag_dr * factor, 0, size - 1),
                    clamp(next_shadow.1 + lag_dc * factor, 0, size - 1),
                );
            }

            let mut next_history = current.state.history.clone();
            next_history.push((dr, dc));
            if next_history.len() > level.lag {
                next_history.remove(0);
            }

            let p_on = on_pad(next_player.0, next_player.1);
            let s_on = on_pad(next_shadow.0, next_shadow.1);
            let open = current.state.open || (p_on && s_on && next_player != next_shadow);

            let next_key = StateKey {
                player: next_player,
                shadow: next_shadow,
                history: next_history,
                open,
            };

            if !seen.contains(&next_key) {
                seen.insert(next_key.clone());
                queue.push_back(QueueItem {
                    state: next_key,
                    steps: current.steps + 1,
                });
            }
        }
    }

    false
}

pub fn generate_shadow_level(seed: u32, mastery: u32) -> ShadowLevel {
    for attempt in 0..12 {
        let candidate = build_shadow_level_candidate(seed.wrapping_add(attempt * 7331), mastery);
        if is_shadow_level_solvable(&candidate) {
            return candidate;
        }
    }
    build_shadow_level_candidate(seed, mastery)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_shadow_level_generation() {
        let level = generate_shadow_level(101, 1);
        assert_eq!(level.size, 5);
        assert_eq!(level.pads.len(), 2);
        assert_eq!(level.lag, 2);
        assert!(is_shadow_level_solvable(&level));
    }

    #[test]
    fn test_shadow_level_mastery_high() {
        let level = generate_shadow_level(202, 3);
        assert_eq!(level.size, 6);
        assert_eq!(level.inverse_tiles.len(), 1);
    }
}
